File: app/src/main/java/cn/archko/pdf/ChooseFileFragmentActivity.java
package cn.archko.pdf;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.MenuItemCompat;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.util.SparseArray;
import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.widget.Toast;

import cx.hell.android.pdfviewpro.APVApplication;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

/**
 * Minimalistic file browser.
 *
 * @author archko
 */
public class ChooseFileFragmentActivity extends FragmentActivity {

    /**
     * Logging tag.
     */
    private final static String TAG = "pdfviewpro";

    public final static String PREF_TAG = "ChooseFileActivity";
    public final static String PREF_HOME = "Home";

    ViewPager mViewPager;
    TabsAdapter mPagerAdapter;
    final String[] titles = new String[2];
    private static final int REQUEST_PERMISSION_CODE = 0x001;

    /**
     * List of {@link SamplePagerItem} which represent this sample's tabs.
     */
    protected List<SamplePagerItem> mTabs = new ArrayList<SamplePagerItem>();
    /**
     * A custom {@link android.support.v4.view.ViewPager} title strip which looks much like Tabs present in Android v4.0 and
     * above, but is designed to give continuous feedback to the user when scrolling.
     */
    protected SlidingTabLayout mSlidingTabLayout;
    private MenuItem searchMenuItem;

    /**
     * This class represents a tab to be displayed by {@link android.support.v4.view.ViewPager} and it's associated
     * {@link cn.archko.pdf.SlidingTabLayout}.
     */
    static class SamplePagerItem {

        private final CharSequence mTitle;
        private final int mIndicatorColor;
        private final int mDividerColor;
        private final Class<?> clss;
        private final Bundle args;

        SamplePagerItem(Class<?> _class, Bundle _args, CharSequence title) {
            clss = _class;
            args = _args;
            mTitle = title;
            mIndicatorColor = APVApplication.getInstance().getResources().getColor(R.color.tab_indicator_selected_color);
            mDividerColor = APVApplication.getInstance().getResources().getColor(R.color.tab_divider_color);
        }

        SamplePagerItem(Class<?> _class, Bundle _args, CharSequence title, int indicatorColor, int dividerColor) {
            clss = _class;
            args = _args;
            mTitle = title;
            mIndicatorColor = indicatorColor;
            mDividerColor = dividerColor;
        }

        /**
         * @return the title which represents this tab. In this sample this is used directly by
         * {@link android.support.v4.view.PagerAdapter#getPageTitle(int)}
         */
        CharSequence getTitle() {
            return mTitle;
        }

        /**
         * @return the color to be used for indicator on the {@link cn.archko.pdf.SlidingTabLayout}
         */
        int getIndicatorColor() {
            return mIndicatorColor;
        }

        /**
         * @return the color to be used for right divider on the {@link cn.archko.pdf.SlidingTabLayout}
         */
        int getDividerColor() {
            return mDividerColor;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.fragment_tabs_pager);

        checkSdcardPermission();
    }

    private void loadView() {
        mViewPager = (ViewPager) findViewById(R.id.pager);

        mSlidingTabLayout = (SlidingTabLayout) findViewById(R.id.sliding_tabs);
        addTab();
        mPagerAdapter = new TabsAdapter(this);
        mViewPager.setOnPageChangeListener(mPagerAdapter);
        mViewPager.setAdapter(mPagerAdapter);

        postSlidingTabLayout();
        // END_INCLUDE (tab_colorizer)
        // END_INCLUDE (setup_slidingtablayout)
        mSlidingTabLayout.setViewPager(mViewPager);
        mSlidingTabLayout.setSelectPageListener(new SlidingTabLayout.SelectPageListener() {
            @Override
            public void updateSubView(int position) {
                ((RefreshableFragment) mPagerAdapter.getItem(position)).update();
            }
        });
    }


    public void checkSdcardPermission() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            // WRITE_EXTERNAL_STORAGE permission has not been granted.

            requestSdcardPermission();
        } else {
            loadView();
        }
    }

    /**
     * Requests the sdcard permission.
     * If the permission has been denied previously, a SnackBar will prompt the user to grant the
     * permission, otherwise it is requested directly.
     */
    private void requestSdcardPermission() {
        Log.i(TAG, "sdcard permission has NOT been granted. Requesting permission.");

        if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            // Provide an additional rationale to the user if the permission was not granted
            // and the user would benefit from additional context for the use of the permission.
            // For example if the user has previously denied the permission.
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    REQUEST_PERMISSION_CODE);
        } else {

            // sdcard permission has not been granted yet. Request it directly.
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    REQUEST_PERMISSION_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == REQUEST_PERMISSION_CODE) {
            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //  权限通过
                //((RefreshableFragment) (mPagerAdapter.getItem(mViewPager.getCurrentItem()))).update();
                loadView();
            } else {
                // 权限拒绝
                Toast.makeText(this, "没有获取sdcard的读取权限", Toast.LENGTH_LONG).show();
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

    }

    protected void postSlidingTabLayout() {
        mSlidingTabLayout.setMatchWidth(true);
        // BEGIN_INCLUDE (tab_colorizer)
        // Set a TabColorizer to customize the indicator and divider colors. Here we just retrieve
        // the tab at the position, and return it's set color
        mSlidingTabLayout.setCustomTabColorizer(new SlidingTabLayout.TabColorizer() {

            @Override
            public int getIndicatorColor(int position) {
//                LLog.d("JSSlidingTabsColorsFragment", "*****getIndicatorColor******"+mTabs.get(position).getTitle());
                return mTabs.get(position).getIndicatorColor();
            }

            @Override
            public int getDividerColor(int position) {
//                LLog.d("JSSlidingTabsColorsFragment", "******getDividerColor*****"+mTabs.get(position).getTitle());
                return mTabs.get(position).getDividerColor();
            }

        });
    }

    protected void addTab() {
        titles[0] = getString(R.string.tab_history);
        titles[1] = getString(R.string.tab_browser);

        String title = titles[0];
        Bundle bundle = new Bundle();
        mTabs.add(new SamplePagerItem(HistoryFragment.class, bundle, title));

        title = titles[1];
        bundle = new Bundle();
        mTabs.add(new SamplePagerItem(BrowserFragment.class, bundle, title));
    }

    @Override
    public void onBackPressed() {
        if (null != mPagerAdapter) {
            BrowserFragment fragment = (BrowserFragment) mPagerAdapter.getItem(mViewPager.getCurrentItem());
            if (fragment.onBackPressed()) {
                return;
            }
        }
        super.onBackPressed();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        boolean flag= super.onCreateOptionsMenu(menu);
        this.searchMenuItem = menu.add(R.string.menu_search);
        MenuItemCompat.setShowAsAction(this.searchMenuItem, MenuItemCompat.SHOW_AS_ACTION_IF_ROOM);
        return flag;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        if (menuItem == this.searchMenuItem) {
            showSearchDialog();
        }
        return super.onOptionsItemSelected(menuItem);
    }


    protected void showSearchDialog() {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        Fragment prev = getSupportFragmentManager().findFragmentByTag("dialog");
        if (prev != null) {
            ft.remove(prev);
        }
        ft.addToBackStack(null);

        // Create and show the dialog.
        SearchFragment fileInfoFragment = new SearchFragment();
        Bundle bundle = new Bundle();
        fileInfoFragment.setArguments(bundle);
        fileInfoFragment.show(ft, "dialog");
    }

    /**
     * The {@link android.support.v4.app.FragmentPagerAdapter} used to display pages in this sample. The individual pages
     * are instances of {@link ContentFragment} which just display three lines of text. Each page is
     * created by the relevant {@link SamplePagerItem} for the requested position.
     * <p/>
     * The important section of this class is the {@link #getPageTitle(int)} method which controls
     * what is displayed in the {@link SlidingTabLayout}.
     */
    public class TabsAdapter extends FragmentPagerAdapter
            implements ViewPager.OnPageChangeListener {
        Handler mHandler = new Handler();
        private final Context mContext;
        private final SparseArray<WeakReference<Fragment>> mFragmentArray = new SparseArray<WeakReference<Fragment>>();
        int position = 0;

        public TabsAdapter(FragmentActivity activity) {
            super(activity.getSupportFragmentManager());
            mContext = activity;
        }

        @Override
        public int getCount() {
            return mTabs.size();
        }

        @Override
        public Fragment getItem(int position) {
            final WeakReference<Fragment> mWeakFragment = mFragmentArray.get(position);
            if (mWeakFragment != null && mWeakFragment.get() != null) {
                return mWeakFragment.get();
            }

            SamplePagerItem info = mTabs.get(position);
            return Fragment.instantiate(mContext, info.clss.getName(), info.args);
        }

        @Override
        public Object instantiateItem(final ViewGroup container, final int position) {
            Log.v(TAG, "instantiateItem:" + position);
            WeakReference<Fragment> mWeakFragment = mFragmentArray.get(position);
            if (mWeakFragment != null && mWeakFragment.get() != null) {
                //mWeakFragment.clear();
                return mWeakFragment.get();
            }

            final Fragment mFragment = (Fragment) super.instantiateItem(container, position);
            mFragmentArray.put(position, new WeakReference<Fragment>(mFragment));
            return mFragment;
        }

        @Override
        public void destroyItem(final ViewGroup container, final int position, final Object object) {
            super.destroyItem(container, position, object);
            final WeakReference<Fragment> mWeakFragment = mFragmentArray.get(position);
            if (mWeakFragment != null) {
                mWeakFragment.clear();
            }
        }

        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        }

        @Override
        public void onPageSelected(int position) {
            mViewPager.setCurrentItem(position);
            if (position == 0) {
                APVApplication apvApplication = APVApplication.getInstance();
                if (apvApplication.hasChanged) {
                    delayUpdate(position);
                    apvApplication.hasChanged = false;
                }
            } else if (position == 2) {
                delayUpdate(position);
            }
        }

        @Override
        public void onPageScrollStateChanged(int state) {
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mTabs.get(position).getTitle();
        }

        void delayUpdate(final int position) {
            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    RefreshableFragment fragment = (RefreshableFragment) getItem(position);
                    fragment.update();
                }
            }, 500L);
        }
    }
}

File: app/src/main/java/cn/archko/pdf/ChooseFileFragmentActivity.kt
package cn.archko.pdf

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.MenuItemCompat
import android.support.v4.view.ViewPager
import android.util.Log
import android.util.SparseArray
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup
import android.widget.Toast
import cx.hell.android.pdfviewpro.APVApplication
import java.lang.ref.WeakReference
import java.util.*

/**
 * Minimalistic file browser.

 * @author archko
 */
open class ChooseFileFragmentActivity : FragmentActivity() {

    internal var mViewPager: ViewPager? = null
    internal var mPagerAdapter: TabsAdapter? = null
    internal val titles = arrayOfNulls<String>(2)

    /**
     * List of [SamplePagerItem] which represent this sample's tabs.
     */
    internal var mTabs: MutableList<SamplePagerItem> = ArrayList()
    /**
     * A custom [android.support.v4.view.ViewPager] title strip which looks much like Tabs present in Android v4.0 and
     * above, but is designed to give continuous feedback to the user when scrolling.
     */
    protected var mSlidingTabLayout: SlidingTabLayout? = null
    private var searchMenuItem: MenuItem? = null

    /**
     * This class represents a tab to be displayed by [android.support.v4.view.ViewPager] and it's associated
     * [cn.archko.pdf.SlidingTabLayout].
     */
    internal class SamplePagerItem {

        /**
         * @return the title which represents this tab. In this sample this is used directly by
         * * [android.support.v4.view.PagerAdapter.getPageTitle]
         */
        val title: CharSequence
        /**
         * @return the color to be used for indicator on the [cn.archko.pdf.SlidingTabLayout]
         */
        val indicatorColor: Int
        /**
         * @return the color to be used for right divider on the [cn.archko.pdf.SlidingTabLayout]
         */
        val dividerColor: Int
        val clss: Class<*>
        val args: Bundle

        constructor(_class: Class<*>, _args: Bundle, title: CharSequence) {
            clss = _class
            args = _args
            this.title = title
            indicatorColor = APVApplication.getInstance().resources.getColor(R.color.tab_indicator_selected_color)
            dividerColor = APVApplication.getInstance().resources.getColor(R.color.tab_divider_color)
        }

        constructor(_class: Class<*>, _args: Bundle, title: CharSequence, indicatorColor: Int, dividerColor: Int) {
            clss = _class
            args = _args
            this.title = title
            this.indicatorColor = indicatorColor
            this.dividerColor = dividerColor
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.fragment_tabs_pager)

        checkSdcardPermission()
    }

    private fun loadView() {
        mViewPager = findViewById<ViewPager>(R.id.pager)

        mSlidingTabLayout = findViewById<SlidingTabLayout>(R.id.sliding_tabs)
        addTab()
        mPagerAdapter = TabsAdapter(this)
        mViewPager!!.setOnPageChangeListener(mPagerAdapter)
        mViewPager!!.adapter = mPagerAdapter

        postSlidingTabLayout()
        // END_INCLUDE (tab_colorizer)
        // END_INCLUDE (setup_slidingtablayout)
        mSlidingTabLayout!!.setViewPager(mViewPager)
        mSlidingTabLayout!!.setSelectPageListener { position -> (mPagerAdapter!!.getItem(position) as RefreshableFragment).update() }
    }


    fun checkSdcardPermission() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            // WRITE_EXTERNAL_STORAGE permission has not been granted.

            requestSdcardPermission()
        } else {
            loadView()
        }
    }

    /**
     * Requests the sdcard permission.
     * If the permission has been denied previously, a SnackBar will prompt the user to grant the
     * permission, otherwise it is requested directly.
     */
    private fun requestSdcardPermission() {
        Log.i(TAG, "sdcard permission has NOT been granted. Requesting permission.")

        if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            // Provide an additional rationale to the user if the permission was not granted
            // and the user would benefit from additional context for the use of the permission.
            // For example if the user has previously denied the permission.
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                    REQUEST_PERMISSION_CODE)
        } else {

            // sdcard permission has not been granted yet. Request it directly.
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                    REQUEST_PERMISSION_CODE)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == REQUEST_PERMISSION_CODE) {
            if (grantResults.size == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //  权限通过
                //((RefreshableFragment) (mPagerAdapter.getItem(mViewPager.getCurrentItem()))).update();
                loadView()
            } else {
                // 权限拒绝
                Toast.makeText(this, "没有获取sdcard的读取权限", Toast.LENGTH_LONG).show()
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

    }

    protected fun postSlidingTabLayout() {
        mSlidingTabLayout!!.setMatchWidth(true)
        // BEGIN_INCLUDE (tab_colorizer)
        // Set a TabColorizer to customize the indicator and divider colors. Here we just retrieve
        // the tab at the position, and return it's set color
        mSlidingTabLayout!!.setCustomTabColorizer(object : SlidingTabLayout.TabColorizer {

            override fun getIndicatorColor(position: Int): Int {
                //                LLog.d("JSSlidingTabsColorsFragment", "*****getIndicatorColor******"+mTabs.get(position).getTitle());
                return mTabs[position].indicatorColor
            }

            override fun getDividerColor(position: Int): Int {
                //                LLog.d("JSSlidingTabsColorsFragment", "******getDividerColor*****"+mTabs.get(position).getTitle());
                return mTabs[position].dividerColor
            }

        })
    }

    protected fun addTab() {
        titles[0] = getString(R.string.tab_history)
        titles[1] = getString(R.string.tab_browser)

        var title = titles[0]
        var bundle = Bundle()
        mTabs.add(SamplePagerItem(HistoryFragment::class.java, bundle, title!!))

        title = titles[1]
        bundle = Bundle()
        mTabs.add(SamplePagerItem(BrowserFragment::class.java, bundle, title!!))
    }

    override fun onBackPressed() {
        if (null != mPagerAdapter) {
            val fragment = mPagerAdapter!!.getItem(mViewPager!!.currentItem) as BrowserFragment
            if (fragment.onBackPressed()) {
                return
            }
        }
        super.onBackPressed()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val flag = super.onCreateOptionsMenu(menu)
        this.searchMenuItem = menu.add(R.string.menu_search)
        MenuItemCompat.setShowAsAction(this.searchMenuItem, MenuItemCompat.SHOW_AS_ACTION_IF_ROOM)
        return flag
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        if (menuItem === this.searchMenuItem) {
            showSearchDialog()
        }
        return super.onOptionsItemSelected(menuItem)
    }


    protected fun showSearchDialog() {
        val ft = supportFragmentManager.beginTransaction()
        val prev = supportFragmentManager.findFragmentByTag("dialog")
        if (prev != null) {
            ft.remove(prev)
        }
        ft.addToBackStack(null)

        // Create and show the dialog.
        val fileInfoFragment = SearchFragment()
        val bundle = Bundle()
        fileInfoFragment.arguments = bundle
        fileInfoFragment.show(ft, "dialog")
    }

    /**
     * The [android.support.v4.app.FragmentPagerAdapter] used to display pages in this sample. The individual pages
     * are instances of [ContentFragment] which just display three lines of text. Each page is
     * created by the relevant [SamplePagerItem] for the requested position.
     *
     *
     * The important section of this class is the [.getPageTitle] method which controls
     * what is displayed in the [SlidingTabLayout].
     */
    inner class TabsAdapter(activity: FragmentActivity) : FragmentPagerAdapter(activity.supportFragmentManager), ViewPager.OnPageChangeListener {
        internal var mHandler = Handler()
        private val mContext: Context
        private val mFragmentArray = SparseArray<WeakReference<Fragment>>()
        internal var position = 0

        init {
            mContext = activity
        }

        override fun getCount(): Int {
            return mTabs.size
        }

        override fun getItem(position: Int): Fragment {
            val mWeakFragment = mFragmentArray.get(position)
            if (mWeakFragment != null && mWeakFragment.get() != null) {
                return mWeakFragment.get()!!
            }

            val info = mTabs[position]
            return Fragment.instantiate(mContext, info.clss.name, info.args)
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            Log.v(TAG, "instantiateItem:" + position)
            val mWeakFragment = mFragmentArray.get(position)
            if (mWeakFragment != null && mWeakFragment.get() != null) {
                //mWeakFragment.clear();
                return mWeakFragment.get()!!
            }

            val mFragment = super.instantiateItem(container, position) as Fragment
            mFragmentArray.put(position, WeakReference(mFragment))
            return mFragment
        }

        override fun destroyItem(container: ViewGroup?, position: Int, `object`: Any) {
            super.destroyItem(container, position, `object`)
            val mWeakFragment = mFragmentArray.get(position)
            mWeakFragment?.clear()
        }

        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

        override fun onPageSelected(position: Int) {
            mViewPager!!.currentItem = position
            /*if (position == 0) {
                APVApplication apvApplication = APVApplication.getInstance();
                if (apvApplication.hasChanged) {
                    delayUpdate(position);
                    apvApplication.hasChanged = false;
                }
            } else if (position == 2) {
                delayUpdate(position);
            }*/
        }

        override fun onPageScrollStateChanged(state: Int) {}

        override fun getPageTitle(position: Int): CharSequence {
            return mTabs[position].title
        }

        internal fun delayUpdate(position: Int) {
            mHandler.postDelayed({
                val fragment = getItem(position) as RefreshableFragment
                fragment.update()
            }, 500L)
        }
    }

    companion object {

        /**
         * Logging tag.
         */
        private val TAG = "pdfviewpro"

        @JvmField
        val PREF_TAG = "ChooseFileActivity"
        @JvmField
        val PREF_HOME = "Home"
        private val REQUEST_PERMISSION_CODE = 0x001
    }
}
